opentelemetry-sdk/src/logs/log_emitter.rs

```diff
@@ -0,0 +1,259 @@
+        self,
+        exporter: T,
+        runtime: R,
+    ) -> Self {
+        let batch = BatchLogProcessor::builder(exporter, runtime).build();
+        self.with_log_processor(batch)
+    }
+
+    /// The `LogProcessor` that this provider should use.
+    pub fn with_log_processor<T: LogProcessor + 'static>(self, processor: T) -> Self {
+        let mut processors = self.processors;
+        processors.push(Box::new(processor));
+
+        Builder { processors, ..self }
+    }
+
+    /// The `Config` that this provider should use.
+    pub fn with_config(self, config: Config) -> Self {
+        Builder { config, ..self }
+    }
+
+    /// Return the clone of sdk provided resource.
+    ///
+    /// See <https://github.com/open-telemetry/opentelemetry-specification/blob/v1.8.0/specification/resource/sdk.md#sdk-provided-resource-attributes>
+    /// for details.
+    pub fn sdk_provided_resource(&self) -> Resource {
+        self.sdk_provided_resource.clone()
+    }
+
+    /// Create a new provider from this configuration.
+    pub fn build(self) -> LogEmitterProvider {
+        let mut config = self.config;
+        config.resource = match config.resource {
+            None => Some(Arc::new(self.sdk_provided_resource)),
+            Some(resource) => {
+                if resource.is_empty() {
+                    None
+                } else {
+                    Some(Arc::new(self.sdk_provided_resource.merge(resource)))
+                }
+            }
+        };
+
+        LogEmitterProvider {
+            inner: Arc::new(LogEmitterProviderInner {
+                processors: self.processors,
+                config,
+            }),
+        }
+    }
+}
+
+#[derive(Debug)]
+/// The object for emitting [`LogRecord`]s.
+///
+/// [`LogRecord`]: crate::log::LogRecord
+pub struct LogEmitter {
+    instrumentation_lib: InstrumentationLibrary,
+    provider: Weak<LogEmitterProviderInner>,
+}
+
+impl LogEmitter {
+    pub(crate) fn new(
+        instrumentation_lib: InstrumentationLibrary,
+        provider: Weak<LogEmitterProviderInner>,
+    ) -> Self {
+        LogEmitter {
+            instrumentation_lib,
+            provider,
+        }
+    }
+
+    /// LogEmitterProvider associated with this tracer.
+    pub fn provider(&self) -> Option<LogEmitterProvider> {
+        self.provider.upgrade().map(LogEmitterProvider::new)
+    }
+
+    /// Instrumentation library information of this tracer.
+    pub fn instrumentation_library(&self) -> &InstrumentationLibrary {
+        &self.instrumentation_lib
+    }
+
+    /// Emit a `LogRecord`.
+    pub fn emit(&self, record: LogRecord) {
+        let provider = match self.provider() {
+            Some(provider) => provider,
+            None => return,
+        };
+
+        let config = provider.config();
+        for processor in provider.log_processors() {
+            let data = LogData {
+                record: record.clone(),
```

Comment:
nit - is this necessary to clone the data, or can we pass it as reference to the processor/exporter. The clone operation is in general extensive.